lz4/stream/_stream.c

```diff
@@ -0,0 +1,1404 @@
+/*
+ * Copyright (c) 2019, Samuel Martin
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of Steeve Morin nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software without
+ *    specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if defined(_WIN32) && defined(_MSC_VER)
+#define inline __inline
+#elif defined(__SUNPRO_C) || defined(__hpux) || defined(_AIX)
+#define inline
+#endif
+
+#include <py3c.h>
+#include <py3c/capsulethunk.h>
+#include <stdlib.h>
+#include <math.h>
+#include <lz4.h>
+#include <lz4hc.h>
+#include <stddef.h>
+#include <stdint.h>
+
+//#define DUMP_CONTEXT
+
+#if (LZ4_VERSION_MAJOR >= 1) && (LZ4_VERSION_MINOR >= 9) && (LZ4_VERSION_RELEASE >= 0)
+#define LZ4_VERSION_GE_1_9_0  1
+#pragma message "Building bindings against LZ4 " LZ4_VERSION_STRING ", using API>=1.9.0: yes"
+#else
+#define LZ4_VERSION_GE_1_9_0  0
+#pragma message "Building bindings against LZ4 " LZ4_VERSION_STRING ", using API>=1.9.0: no"
+#endif
+
```

Comment:
In general I prefer runtime detection of the LZ4 version rather than compile time: that way, if e build against system installed LZ4 (as distros tend to do), things shouldn't break when the distro updates the system LZ4. See `_frame.c` for an example of using `LZ4_versionNumber()`